src/nbl/asset/utils/CHLSLCompiler.cpp

```diff
@@ -25,22 +25,712 @@
+    {
         ComPtr<IDxcUtils> m_dxcUtils;
         ComPtr<IDxcCompiler3> m_dxcCompiler;
     };
+
+
+    // for including builtins 
+    struct load_file_or_builtin_to_string
+    {
+        template <typename IterContextT>
+        class inner
+        {
+        public:
+            template <typename PositionT>
+            static void init_iterators(IterContextT& iter_ctx, PositionT const& act_pos, boost::wave::language_support language)
+            {
+                using iterator_type = typename IterContextT::iterator_type;
+
+                std::string filepath(iter_ctx.filename.begin(), iter_ctx.filename.end());
+                auto inclFinder = iter_ctx.ctx.get_hooks().m_includeFinder;
+                if (inclFinder) 
+                {
+                    std::optional<std::string> result;
+                    system::path requestingSourceDir(iter_ctx.ctx.get_current_directory().string());
+                    if (iter_ctx.type == IterContextT::base_type::file_type::system_header) // is it a sys include (#include <...>)?
+                        result = inclFinder->getIncludeStandard(requestingSourceDir, filepath);
+                    else // regular #include "..."
+                        result = inclFinder->getIncludeRelative(requestingSourceDir, filepath);
+
+                    if (!result)
+                        BOOST_WAVE_THROW_CTX(iter_ctx.ctx, boost::wave::preprocess_exception,
+                            bad_include_file, iter_ctx.filename.c_str(), act_pos);
+                    auto& res_str = *result;
+                    iter_ctx.instring = res_str;
+                }
+                iter_ctx.first = iterator_type(
+                    iter_ctx.instring.begin(), iter_ctx.instring.end(),
+                    PositionT(iter_ctx.filename), language);
+                iter_ctx.last = iterator_type();
+            }
+
+        private:
+            std::string instring;
+        };
+    };
+
+
+    struct custom_preprocessing_hooks : public boost::wave::context_policies::default_preprocessing_hooks
+    {
+
+        custom_preprocessing_hooks(const IShaderCompiler::SPreprocessorOptions& _preprocessOptions) 
+            : m_includeFinder(_preprocessOptions.includeFinder), m_logger(_preprocessOptions.logger), m_pragmaStage(IShader::ESS_UNKNOWN) {}
+
+        const IShaderCompiler::CIncludeFinder* m_includeFinder;
+        system::logger_opt_ptr m_logger;
+        IShader::E_SHADER_STAGE m_pragmaStage;
+
+
+        template <typename ContextT>
+        bool locate_include_file(ContextT& ctx, std::string& file_path, bool is_system, char const* current_name, std::string& dir_path, std::string& native_name) 
+        {
+            //on builtin return true
+            dir_path = ctx.get_current_directory().string();
+            std::optional<std::string> result;
+            if (is_system) {
+                result = m_includeFinder->getIncludeStandard(dir_path, file_path);
+                dir_path = "";
```

Comment:
what's this?

Reply:
Where's all this code? https://github.com/Devsh-Graphics-Programming/Nabla/pull/608#discussion_r1399297717

Reply:
`dir_path` appears to be the directory path of THE NEW INCLUDE 
NOT THE REQUESTING INCLUDE!

I asked for some research!
https://github.com/Devsh-Graphics-Programming/Nabla/pull/608#discussion_r1399310698